crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -147,80 +60,185 @@ impl SystemStage {
+                          conflicts, systems, ..
+                      }| {
+                    let conflicts = conflicts
+                        .iter()
+                        .map(|id| world.components().get_info(*id).unwrap().name().to_owned())
+                        .collect();
+                    let mut system_names: Vec<_> = systems
+                        .iter()
+                        .map(|&SystemIndex(i)| container[i].name().to_string())
+                        .collect();
+                    system_names.sort();
+                    SystemOrderAmbiguity {
+                        segment,
+                        conflicts,
+                        system_names,
+                    }
+                },
+            )
+        }
+
+        find_ambiguities_in(SystemStageSegment::Parallel, &self.parallel, world)
+            .chain(find_ambiguities_in(
+                SystemStageSegment::ExclusiveAtStart,
+                &self.exclusive_at_start,
+                world,
+            ))
+            .chain(find_ambiguities_in(
+                SystemStageSegment::ExclusiveBeforeCommands,
+                &self.exclusive_before_commands,
+                world,
+            ))
+            .chain(find_ambiguities_in(
+                SystemStageSegment::ExclusiveAtEnd,
+                &self.exclusive_at_end,
+                world,
+            ))
+            .collect()
     }
 
     /// Returns the number of system order ambiguities between systems in this stage.
     ///
     /// The result may be incorrect if this stage has not been initialized with `world`.
     #[cfg(test)]
     fn ambiguity_count(&self, world: &World) -> usize {
-        self.ambiguities(world).len()
+        fn binomial_coefficient(n: usize, k: usize) -> usize {
+            (0..k).fold(1, |x, i| x * (n - i) / (i + 1))
+        }
+
+        self.ambiguities(world)
+            .iter()
+            .map(|a| binomial_coefficient(a.system_names.len(), 2))
+            .sum()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct SystemOrderAmbiguity {
+    segment: SystemStageSegment,
+    conflicts: Vec<String>,
+    system_names: Vec<String>,
+}
+
+/// Which part of a [`SystemStage`] was a [`SystemOrderAmbiguity`] detected in?
+#[derive(Debug, PartialEq, Eq, Clone, Copy, PartialOrd, Ord, Hash)]
+enum SystemStageSegment {
+    Parallel,
+    ExclusiveAtStart,
+    ExclusiveBeforeCommands,
+    ExclusiveAtEnd,
+}
+
+impl SystemStageSegment {
+    pub fn desc(&self) -> &'static str {
+        match self {
+            SystemStageSegment::Parallel => "Parallel systems",
+            SystemStageSegment::ExclusiveAtStart => "Exclusive systems at start of stage",
+            SystemStageSegment::ExclusiveBeforeCommands => {
+                "Exclusive systems before commands of stage"
+            }
+            SystemStageSegment::ExclusiveAtEnd => "Exclusive systems at end of stage",
+        }
+    }
+}
```

Comment:
Wondering if it should implement `Display` instead. Anyway, since this is just a code move, I believe the change should be applied separately.

Reply:
Good point, this is just churn. I'll move it back.